strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/GroupRepositoryProvider.java

```diff
@@ -106,57 +115,96 @@ protected RepositoryPath resolvePathTraversal(RepositoryPath repositoryPath) thr
             {
                 continue;
             }
-            
-            RepositoryPath result = repositoryPathResolver.resolve(r, repositoryPath);
+
+            result = repositoryPathResolver.resolve(r, repositoryPath);
             if (artifactRoutingRulesChecker.isDenied(groupRepository.getId(), result))
             {
+                result = null;
                 continue;
             }
-            
-            result = resolvePathFromGroupMemberOrTraverse(result);
-            if (result == null)
+
+            final RepositoryPath selfResult = result;
+            actions.add(() -> resolvePathFromGroupMemberOrTraverse(selfResult, firstMatch));
+
+            if (firstMatch)
             {
-                continue;
+                result = invokeResolvePathActions(actions);
+                if (result == null)
+                {
+                    continue;
+                }
+                logger.debug(String.format("Located artifact: [%s]", result));
+                return result;
+            }
+        }
+
+        if (!firstMatch)
+        {
+            result = invokeResolvePathActions(actions);
+        }
+
+        return result;
+    }
+
+    private RepositoryPath invokeResolvePathActions(List<Callable<RepositoryPath>> resolvePathActions)
+    {
+        List<Future<RepositoryPath>> answers;
+        try
+        {
+            // TODO invokeAll timeout interrupt ?
+            answers = executorService.invokeAll(resolvePathActions);
```

Comment:
I think we can use `Collection.parallelStream()` here instead of custom executor.

Reply:
Then I'd have to use custom `ForkJoinPool` because I wouldn't want to use the default one, because by default it is configured to use as many threads as there are cores in the system.

What's the benefit of choosing `Collection.parallelStream()` ?

Reply:
> What's the benefit of choosing Collection.parallelStream() ?

As I understand `ForkJoinPool.commonPool()`  use same parallelism as you calculate manually for `@CommonExecutorService` (number of cores minus one), but it also use caller thread as worker thread. 
So I think `@CommonExecutorService` redundant and less productive then default `ForkJoinPool`.

Reply:
> As I understand ForkJoinPool.commonPool() use same parallelism as you calculate manually for @CommonExecutorService (number of cores minus one), 

Looks like you're right, thanks for this info :-)

> but it also use caller thread as worker thread.

Where did you find it ?

> So I think @CommonExecutorService redundant

Ok, I'm convinced now :-)

>  and less productive then default ForkJoinPool.

Hmm.. why so ? :-)

Reply:
you can check it with this code:
```
System.out.println(String.format("This is the caller thread: [%s]", Thread.currentThread().getName()));
        IntStream.range(0, Runtime.getRuntime().availableProcessors())
                 .parallel()
                 .forEach(i -> {
                     try
                     {
                         Thread.sleep(100);
                     }
                     catch (InterruptedException e)
                     {
                         e.printStackTrace();
                     }
                     System.out.println(Thread.currentThread().getName());
                 });
```

Reply:
> because it utilize caller thread as well

Ok, so summarizing, `Collection.parallelStream()` uses #`total number of available cores` threads = (`number of cores minus one` (from `ForkJoinPool.commonPool()`)+ `the caller thread`.

This whole conversation led me to the conclusion that we shouldn't use `common` pools all over the code base in our specific environment. Using the same `ForkJoinPool.commonPool()` (or `@CommonExecutorService` in many places may lead to block other parts of the code (other streams) using the same `common` pool. 

I'm gonna use specific `ExecutorService`/`thread pool` for `GroupRepositoryProvider`.

Reply:
> Using the same ForkJoinPool.commonPool() (or @CommonExecutorService in many places may lead to block other parts of the code (other streams) using the same common pool.

with `ForkJoinPool.commonPool()` there is no way to be completely blocked because it will always have at least 1 available thread (which is caller therad) .